### ISSUE TITLE
Elastic Agent should support a restart and uninstall action

### ISSUE THREAD

**Issue by jlind23** [elastic/elastic-agent#3367]
In order to give our users the ability to remotely restart an Agent, the Elastic Agent should support a restart action.
We have two path forward:
This should have the exact same behaviour as the `elastic-agent restart` and `elastic-agent uninstall` command.

**Comment by jlind23** [comment]
@cmacknz @pierrehilbert I believe this is the first piece to deliver before working on https://github.com/elastic/kibana/issues/144585

**Comment by pierrehilbert** [comment]
From my point of view, we "just" need to add a new handler in the same way we did it for remote diag or upgrades.
What is the priority for this new feature compare to the other we have?

**Comment by blakerouse** [comment]
The hardest part of adding the action is how to handle the ACK. Do we ACK before we restart or after we restart. If we do it after then we need to handle that, if we do it before we need to ensure the ACK is received by Fleet Server before we perform the restart.

**Comment by cmacknz** [comment]
We should acknowledge only after the restart has completed. This is basically the same problem as acknowledging an upgrade, and the easiest solution to implement without edge cases will be to report the state in as part of the periodic check in rather than trying to guarantee delivery of a single acknowledgment across restarts.

There are probably a few ways to accomplish this, off the top of my head including any kind of monotonic number in the check in that resets on a restart is one way for Fleet to detect this without an explicit acknowledgement. A counter for the number of restarts performed, the accumulated time since the last restart / since the agent process started, etc.

We could also persist the acknowledgement to disk and just retry it infinitely until it gets through, but this will have a lot more edge cases to test.

**Comment by nchaulet** [comment]
Hi I am working on the tech definition for the kibana part https://github.com/elastic/kibana/issues/144585 and I am wondering what we will have to send to the agent in that case? does a new action type something like `RESTART` will work for you?

**Comment by cmacknz** [comment]
Not defined yet, a restart action is one way, but we could also define it as a special case of the upgrade action so we can reuse the state reporting and acknowledgement mechanisms from that.

**Comment by cmacknz** [comment]
The hardest parts of a restart action is that the ack should technically happen after the restart, which requires that state be tracked outside of the agent process, which is exactly what our upgrade state reporting is built to do. Also, "restart yourself" is already a step that exists in the upgrade state machine. Perhaps there is a separate restart action that is just an alias for some part of this existing process.

I don't want to over specify this before someone else has a chance to think more in depth about this, but it feels like we can reuse a lot of the existing upgrade machinery to make sure this works properly at an implementation level. We may not actually want to recycle the upgrade action to trigger this.

**Comment by nimarezainia** [comment]
How about our audit logs? we would want to see in the audit logs or the activity panel that the agent was "explicitly" restarted. I believe for that we need a RESTART action as a separate action sent to the agent. As Craig mentions, the RESTART action should invoke the relevant portion of the upgrade process. It's battle tested and has all the necessary state machine states and reporting.

The persona issuing the restart may be different to the one that is doing the upgrade. Upgrade happens at scale but in reality restart may be singular - some one trying to troubleshoot for example. So separate sessions would be issuing these commands. We need to ensure there's a  paper-trail and these actions are distinguished from one another.

> Trying to understand a little more the need here, what information could we want to report after a restart other than acknowledging the restart as been done? could there by some restarts failure?

Almost always the agent is already in a failure state when the user issues a restart (as a desperate last effort move to restore), without having resolved the underlying issue causing the agent to be unhealthy. Chances are that after the restart agent may end up in the same unhealthy/failed state.

**Comment by nchaulet** [comment]
> I don't want to over specify this before someone else has a chance to think more in depth about this, but it feels like we can reuse a lot of the existing upgrade machinery to make sure this works properly at an implementation level. We may not actually want to recycle the upgrade action to trigger this.

👍 who could take a look more in depth about this? from what I understand the implementation could be something like:
*  Fleet UI send a `RESTART` action
* this is handled by the agent with:
 *  audit log about the restart
  * use the same code path as upgrade for restart
  * ack after restart

**Comment by cmacknz** [comment]
@jlind23 is the best person to assign one of the control plane engineers to help refine this.

**Comment by jlind23** [comment]
@blakerouse once you have some spare cycles would you mind helping Nicolas on the request above?

**Comment by cmacknz** [comment]
We probably need to give the UI and user some indicator that we did in fact restart: https://github.com/elastic/kibana/issues/144585#issuecomment-2605713169

We don't rely on the upgrade ack being reliable any more in https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/application/managed_mode.go#L145, we will need to add testing to make sure this ack does eventually get received even if Fleet is temporarily unreachable.

**Comment by nchaulet** [comment]
I guess we could have a `restart_details` passed during checkin with the `restarted_at` information, I am just wondering if it's not adding too much complexity for a feature that is probably only needed for a few issues no?

**Comment by cmacknz** [comment]
Yeah restart_details is too much for this, this will happen so fast it will never be observable in checkins. 

As long as we can guarantee the ack actually gets sent it works to communicate that the restart happened and we can indicate that to the user in some way.

**Comment by nchaulet** [comment]
@cmacknz @blakerouse so what will work is either the agent ack and we know we have a successful restart, or the action is expired? it is the behavior we have for most of the actions like requesting diagnostics  

<img width="800" alt="Image" src="https://github.com/user-attachments/assets/84140515-eda0-48fa-9ad5-785d93cbca5a" />

**Comment by blakerouse** [comment]
@nchaulet Yes I believe checking if it expired is a good way to know if it didn't restart.

**Comment by nimarezainia** [comment]
@jlind23 @ebeahan Could we have this assigned, I believe that it still needs an RFC and we punted it from sp42.

**Comment by nimarezainia** [comment]
@ebeahan @kaanyalti could you please ensure that the implementation issues as the result of the RFC are added to this issue. thank you

**Comment by ebeahan** [comment]
~After more discussion with Control Plane team and @cmacknz, best to improve action system as described in https://github.com/elastic/ingest-dev/issues/1889 before implementing additional actions.~

Update 2026-08-21: no longer taking this approach.

**Comment by nimarezainia** [comment]
@ebeahan we need to have this addressed soon. I know that we lumped it with https://github.com/elastic/ingest-dev/issues/1889 but I am wondering if the restart action can be implemented before having to redesign the whole action system.

**Comment by cmacknz** [comment]
> if the restart action can be implemented before having to redesign the whole action system.

This should be possible, yes.

**Comment by ebeahan** [comment]
we're going to revisit this work in the near term without redesigning the entire action system first.

**Comment by ebeahan** [comment]
We'll also be prioritizing the `RESTART` action first followed by the `UNINSTALL`

**Comment by rdner** [comment]
I've started looking into this.

Do I understand correctly that we're taking the approach described by Blake in https://github.com/elastic/elastic-agent/issues/3367#issuecomment-2604815722 ?

I don't see any blockers from reading through this thread.

**Comment by nimarezainia** [comment]
> I've started looking into this.
> 
> Do I understand correctly that we're taking the approach described by Blake in [#3367 (comment)](https://github.com/elastic/elastic-agent/issues/3367#issuecomment-2604815722) ?
> 
> I don't see any blockers from reading through this thread.

thanks @rdner - yes I believe that is right. Restart and then ACk the restart.

**Comment by rdner** [comment]
The restart action support in Elastic Agent is merged https://github.com/elastic/elastic-agent/pull/16332

Now we can add support in the Fleet UI.

I'll keep working on the uninstall action next.

**Comment by juliaElastic** [comment]
@rdner Does the restart action need changes in Fleet Server? There is an open issue https://github.com/elastic/fleet-server/issues/2523, but it's not clear from the description or comments if there are any changes needed there.

Probably we need to add the RESTART action type and schema to fleet-server.

**Comment by ebeahan** [comment]
@juliaElastic Yes that's correct. I think the new actions need added to the schema and as valid actions in the check-in handler, but I'll let @rdner confirm.

**Comment by rdner** [comment]
> Probably we need to add the RESTART action type and schema to fleet-server.

@juliaElastic exactly, it's currently hard-coded with a TODO in the agent https://github.com/elastic/elastic-agent/blob/b2dcfea10e2adb4a47a00bc8929837f91a4b55eb/pkg/fleetapi/action.go#L32-L36

We also need to make sure that the restart action is signed.

**Comment by rdner** [comment]
These 2 PRs should cover the necessary work in Fleet Server:

* UNINSTALL action https://github.com/elastic/fleet-server/pull/7895 (related to a draft PR [here](https://github.com/elastic/elastic-agent/pull/16540))
* RESTART action https://github.com/elastic/fleet-server/pull/7894 (follow up to [this PR](https://github.com/elastic/elastic-agent/pull/16332))
